modules/kinesis/src/it/scala/com/snowplowanalytics/snowplow/enrich/kinesis/EnrichKinesisSpec.scala

```diff
@@ -0,0 +1,110 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.enrich.kinesis
+
+import scala.concurrent.duration._
+
+import cats.effect.{IO, Resource}
+
+import retry.syntax.all._
+import retry.RetryPolicies
+
+import cats.implicits._
+
+import cats.effect.testing.specs2.CatsIO
+
+import org.testcontainers.containers.{GenericContainer => JGenericContainer}
+
+import org.specs2.mutable.Specification
+import org.specs2.specification.AfterAll
+
+class EnrichKinesisSpec extends Specification with AfterAll with CatsIO {
+
+  override protected val Timeout = 10.minutes
+
+  val retryPolicy =
+    RetryPolicies.limitRetriesByCumulativeDelay(5.minutes, RetryPolicies.capDelay[IO](2.second, RetryPolicies.fullJitter[IO](1.second)))
+
+  def afterAll: Unit = Containers.localstack.stop()
+
+  "enrich-kinesis" should {
+    "initialize with the minimal config" in {
+      Resource.make[IO, JGenericContainer[_]](
+        IO(Containers.enrich(configPath = "config/config.kinesis.minimal.hocon", loggerName = "minimal"))
+      )(
+        e => IO(e.stop())
+      ).use { e =>
+        IO(e.getLogs must contain("Running Enrich"))
+      }
+    }
+
+    "emit the correct number of enriched events and bad rows" in {
+      import count.Helpers._
+
+      val nbGood = 100l
+      val nbBad = 10l
+
+      Resource.make[IO, JGenericContainer[_]](
+        IO(
+          Containers.enrich(
+            configPath = "modules/kinesis/src/it/resources/enrich/count.hocon",
+            loggerName = "count",
+            network = Some(Containers.localstackNetwork)
+          )
+        )
+      )(
+        e => IO(e.stop())
+      ).use { _ =>
+        for {
+          aggregates <- mkResources[IO](nbGood, nbBad, localstackPort = Containers.localStackPort).use { r =>
+            r.readOutput.concurrently(r.writeInput).compile.toList
+          }
+          good = aggregates.collect { case OutputRow.Good(e) => e}
+          bad = aggregates.collect { case OutputRow.Bad(b) => b}
+          _ <- IO(println(s"Bad rows:\n${bad.map(_.compact).mkString("\n")}"))
+        } yield {
+          good.size.toLong === nbGood
+          bad.size.toLong === nbBad
+        }
+      }
+    }
+
+    "shutdown when it receives a SIGTERM" in {
+      Resource.make[IO, JGenericContainer[_]](
+        IO(
+          Containers.enrich(
+            configPath = "modules/kinesis/src/it/resources/enrich/stop.hocon",
+            loggerName = "stop",
+            network = Some(Containers.localstackNetwork)
+          )
+        )
+      )(
+        e => IO(e.stop()) // in case SIGTERM failed to stop the container
+      ).use { enrich =>
+        for {
+          _ <- IO.sleep(20.seconds) // give Enrich more time to initialize (Kinesis)
```

Comment:
What about that idea of replacing sleeping with waiting for a specific log? Is it doable?

Reply:
Can we add a new log? I think it would be nice to have sth like `Enrich initialized properly, ready to process some data`

Reply:
Yes we can do that, but at that point Kinesis SDK won't be fully initialized, as it takes some time to initialize KCL with DynamoDB, listing the shards, syncing etc. and I'd like as much of this to have happened when I send the `SIGTERM`, so that they are the more things that could possibly fail to exit, and we catch it in the tests.

Reply:
I used `metrics.good` that comes later in the initialization